File: 프로그래머스/PCCP 기출문제/2번/석유 시추.py
visit = []
verticals = []
m = {}
numbers = {}

def BFS(n, i, j):
    global visit, verticals, m, numbers

    q = [[i, j]]
    visit[i][j] = n
    numbers[j].add(n)
    cnt = 1

    di = [-1, 0, 1, 0]
    dj = [0, 1, 0, -1]

    while len(q) > 0:
        tmp_i,tmp_j = q[0][0],q[0][1]
        q.pop(0)

        for i in range(4):
            move_i,move_j = tmp_i+di[i],tmp_j+dj[i]

            if move_i<0 or move_i>=len(visit): continue
            if move_j<0 or move_j>=len(visit[0]): continue

            if visit[move_i][move_j] == 1:
                q.append([move_i, move_j])
                visit[move_i][move_j] = n
                numbers[move_j].add(n)
                cnt += 1

    m[n] = cnt
    return

def solution(land):
    global visit, verticals, m, numbers

    visit = land.copy()
    for i in range(len(land[0])):
        numbers[i] = set()
    n = 2

    for i in range(len(land)):
        for j in range(len(land[0])):
            if visit[i][j] == 1:
                BFS(n, i, j)
                n += 1

    verticals = [0 for i in range(len(land[0]))]
    for i in range(len(land[0])):
        for num in numbers[i]:
            verticals[i] += m[num]

    return max(verticals)

File: 프로그래머스/PCCP 기출문제/2번/석유 시추.cpp
#include <string>
#include <vector>
#include <algorithm>
#include <queue>
#include <map>
#include <set>
using namespace std;

vector<vector<int>> visit;   // BFS visit
vector<int> verticals;   // 열마다 시추 양
map<int, int> m;    // <시추 번호, 시추 양>
map<int, set<int>> numbers; // <열 번호, 시추 번호>

void BFS(int n, int i, int j){
    queue<pair<int, int>> q;
    q.push(make_pair(i, j));
    visit[i][j] = n;
    numbers[j].insert(n);
    int cnt = 1;
    
    int di[4] = {-1, 0, 1, 0};
    int dj[4] = {0, 1, 0, -1};
    
    while(q.size()){
        int tmp_i = q.front().first;
        int tmp_j = q.front().second;
        q.pop();
        
        for(int i=0; i<4; i++){
            int move_i = tmp_i+di[i];
            int move_j = tmp_j+dj[i];
            
            if(move_i<0 || move_i>=visit.size()) continue;
            if(move_j<0 || move_j>=visit[0].size()) continue;
            
            if(visit[move_i][move_j] == 1){
                q.push(make_pair(move_i, move_j));
                visit[move_i][move_j] = n;
                numbers[move_j].insert(n);
                cnt++;
            }
        }
    }
    
    m[n] = cnt;
    return;
}

int solution(vector<vector<int>> land) {
    visit = land;
    int n = 2;
    
    for(int i=0; i<land.size(); i++){
        for(int j=0; j<land[0].size(); j++){
            if(visit[i][j] == 1){
                BFS(n, i, j);
                n++;
            }
        }
    }
    
    vector<int> verticals(land[0].size(), 0);   // 열마다 시추 양
    for(int i=0; i<land[0].size(); i++){
        set<int>::iterator it;
        for(it=numbers[i].begin(); it!=numbers[i].end(); it++)
            verticals[i] += m[*it];
    }
    
    return *max_element(verticals.begin(), verticals.end());
}
